ENH Add inverse_transform to random projection transformers

ageron opened the pull request:
#### Reference Issues/PRs

Fixes #21687

#### What does this implement/fix? Explain your changes.

Adds a `fit_inverse_transform` parameter to all transformers in  the `sklearn.random_projection` module: `GaussianRandomProjection` and `SparseRandomProjection`. When set to `True`, the pseudo-inverse of the components is computed during `fit()` and stored in `components_pinv_`, and `inverse_transform()` becomes available.

#### Any other comments?

Using the pseudo-inverse makes sense to me, and it seems to work fine, in the sense that `rnd_proj.transform(rnd_proj.inverse_transform(X))` equals `X`. However, this implementation uses `scipy.linalg.pinv()`, which scales very poorly to large datasets, which is a major use case for random projections. Perhaps it would make sense to use another approach if we detect that the `components_` array is large?

And perhaps there's a mathematical way to generate both a random matrix and its inverse more efficiently?

For the `SparseRandomProjection` transformer, computing the pseudo-inverse breaks sparsity. Perhaps there's a way to generate a sparse matrix that is "close enough" rather than using the pseudo-inverse?

In short: if there's a Random Projection expert in the room, please speak up!

That said, it seems to work fine now, so performance improvements could be pushed in follow-up PRs.

ageron committed:
Add inverse_transform to random projection transformers

ageron committed:
Add PR number. This PR fixes #21687

ageron commented:
The test failure in `sklearn/tests/test_calibration.py` looks unrelated to this PR.

ageron committed:
Test the inverse transform's output shape

ageron committed:
Add missing newline in test

ageron commented:
Acccording to [this SO answer](https://stackoverflow.com/a/13265930/38626):

> `numpy.linalg.pinv()` approximates the Moore-Penrose pseudo inverse using an SVD (the LAPACK method `dgesdd` to be precise), whereas `scipy.linalg.pinv()` solves a model linear system in the least squares sense to approximate the pseudo inverse (using `dgelss`).

They perform differently, both in terms of speed and memory usage, depending on the size of the input. So perhaps a follow-up PR should give the user the option to select the algorithm to use?

adrinjalali commented:
This looks pretty good an clean, thanks @ageron . The only thing I'd add is a section to the relevant user guide (the relevant .rst) to explain the inverse transform, and how it's calculated, and how it's enabled.

glemaitre commented on sklearn/tests/test_random_projection.py:
```diff
@@ -373,3 +373,31 @@ def test_random_projection_feature_names_out(random_projection_cls):
+
+###############################################################################
+# tests on inverse transform
+###############################################################################
```
We would not put this comment.

ogrisel commented:
> numpy.linalg.pinv() approximates the Moore-Penrose pseudo inverse using an SVD (the LAPACK method dgesdd to be precise), whereas scipy.linalg.pinv() solves a model linear system in the least squares sense to approximate the pseudo inverse (using dgelss).

Are you sure this is still valid with recent version of scipy? https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.pinv.html

It's not possible to select the LAPACK driver though.

glemaitre commented on sklearn/random_projection.py:
```diff
@@ -395,6 +403,12 @@ def fit(self, X, y=None):
+            components = self.components_
+            if sp.issparse(components):
+                components = components.toarray()
+            self.components_pinv_ = linalg.pinv(components, check_finite=False)
```
You mentioned that `pinv` does not scale with large matrices. I have 2 questions:

- Should we use `pinv` or `pinv2` depending on the shape of the `X`?
- Is there a way to approximate the inverse? I am recalling the following [PR](https://github.com/scikit-learn/scikit-learn/pull/19971) where we wanted to do so with an approximation in Nystroem (I did not look at the PR thought). I don't know if we could have a trick here to have such an approximation here?

ogrisel replied:
I think it's useful to expose publicly. Maybe we could change the name to `inverse_components_` though to make the name less dependent on an implementation detail.

ogrisel replied:
> Should we use pinv or pinv2 depending on the shape of the X?

pinv2 is deprecated.

ogrisel commented on sklearn/tests/test_random_projection.py:
```diff
@@ -373,3 +373,31 @@ def test_random_projection_feature_names_out(random_projection_cls):
+    assert projected_back.shape == data.shape
+    projected_again = random_projection.transform(projected_back)
+    assert_array_almost_equal(projected, projected_again)
+    assert random_projection.components_pinv_.shape == (n_features, 2)
```
```suggestion
    assert random_projection.components_pinv_.shape == (n_features, 2)
    assert_all_close(
        random_projection.components_ @ random_projection.components_pinv_,
        np.eye(random_projection.n_components)
    )
```

ogrisel reviewed (commented):
Thanks for the PR, it's an interesting contrib. However I have the following concerns + some suggestions:

ageron commented on sklearn/random_projection.py:
```diff
@@ -395,6 +403,12 @@ def fit(self, X, y=None):
+        if self.fit_inverse_transform:
+            components = self.components_
+            if sp.issparse(components):
+                components = components.toarray()
```
I understand. I think I'll go for the second option, and just add fit_inverse_transform to `GaussianRandomProjection` for now.

ageron replied:
Actually, scratch that, I think I'll have a go at using a local `pinv`.

ageron committed:
Add doc on inverse transform in user guide, rename components_pinv_ to inverse_components_, and let inverse_transform() accept sparse arrays

ageron commented:
Thanks @adrinjalali , @glemaitre , and @ogrisel for taking such a thorough look at this PR, and for the interesting feedback.

@ogrisel , I tested both `np.linalg.pinv()` and `scipy.linalg.pinv()` on an array of shape [5_000, 20_000], and SciPy took 25.5 minutes while NumPy took only 11.6 minutes (on Colab). So NumPy's implementation was twice as fast! SciPy uses its `decomp_svd()` function, while NumPy uses its `svd()` function. I haven't looked further yet, so I'm still not sure why NumPy's implementation is twice as fast as SciPy's.

I'll work on the `pinv()` implementation for sparse arrays soon, but in the meantime I made the rest of the changes you suggested, including:

* Added a section in the User Guide in `random_projection.rst`
* Renamed `components_pinv_` to `inverse_components_`. I didn't like the name either!
* Added a note to say that even if `X` is sparse, `X_original` is dense, which may use a lot of RAM.
* Fixed: `X : {array-like, sparse matrix}` **of** `shape (n_samples, n_components)`
* `fit_transform()` now accepts both dense and sparse arrays
* Added testing for both dense and sparse inputs in `test_random_projection.py`
* Removed the `#########` header in `test_random_projection.py`
* Added the test that `random_projection.components_ @ random_projection.inverse_components_` is the identity matrix

ageron commented on sklearn/random_projection.py:
```diff
@@ -431,6 +445,26 @@ def _n_features_out(self):
+        """Project data back to its original space.
+
+        Returns an array X_original whose transform would be X.
+
```
I added "Note that even if `X` is sparse, `X_original` is dense: this may use a lot of RAM."

ogrisel replied:
Maybe we would worth adding a note that users interested in recovering a sparse `X_original` that matches the sparsity level of `X_train` could use `SparseCoder(transform_algorithm="omp")` with `transform_n_nonzero_coefs` set to `X_train.nnz / X_train.shape[0]`.

https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.SparseCoder.html

However I have no idea how tractable this approach is on realistic dataset sizes.

ogrisel replied:
Edit: I realize that `SparseCoder` will always return dense arrays so memory usage remain challenging. It would be possible to convert to sparse matrices in chunks and then concatenate the chunks at the end to avoid blowing up the memory but this would be much more involved.

Maybe it's too complicated and we should not mention `SparseCoder` at all in retrospect.

ogrisel replied:
Still maybe we could store a private:

```python
    self._sparse_X_train = sp.issparse(X)
```

at fit time and then issue a warning in `inverse_transform` to warn about potentially large memory usage in this case prior to calling the projection.

ageron commented on sklearn/tests/test_random_projection.py:
```diff
@@ -373,3 +373,31 @@ def test_random_projection_feature_names_out(random_projection_cls):
+    projected_back = random_projection.inverse_transform(projected)
+    assert projected_back.shape == data.shape
+    projected_again = random_projection.transform(projected_back)
+    assert_array_almost_equal(projected, projected_again)
```
When using `SparseRandomProjection` with sparse `data`, the `components_` matrix is sparse and `projected` is sparse as well. But the `inverse_components_` is always dense, so `projected_back` is always dense, and therefore `projected_again` is also always dense. This is why I added the two lines:

```python
if hasattr(projected, "toarray"):
    projected = projected.toarray()
```

ageron commented:
Implementing `pinv` is trickier than I expected: `scipy.sparse.linalg.svds` expects a `k` argument to specify the number of components, and it must be: 1 ≤ k < min(A.shape)
So it's not possible to use it to get the same result as `scipy.linalg.svd`, since that would require k = min(A.shape).

On a totally different path, it seems possible to [generate randomly a matrix and its inverse efficiently](https://cs.stackexchange.com/a/111738/79057), so perhaps there's a way to generate the components and their pseudo-inverse jointly and efficiently?

Alternatively, depending on the application, it may be preferable to use another random projection to project back to the original space?

I feel like I just escaped Dunning Kruger's mount stupid, and I've now entered the valley of despair: I guess that's progress. What are your thoughts?

ageron committed:
Use scipy's svds to compute pinv when components_ is sparse

ageron committed:
Remove warning about components converted to dense array

ageron commented:
Ok, `SparseRandomProjection.fit()` now computes the pseudo-inverse (when `fit_inverse_transform` is set to True) without converting the `components_` to a dense array.

The pseudo-inverse implementation is based on `scipy.sparse.linalg.svds()`, as you suggested @ogrisel .  I had to call it twice, though: once for the first half of the components, and once for the second half. That's because `scipy.sparse.linalg.svds()` does not support getting all the components at once.

The rest of the pseudo-inverse implementation was taken in large part from `scipy.linalg.pinv()`, but I simplified it a lot, removing every parameter except the sparse matrix. I just wanted to keep things as basic as possible, since this is meant to be a temporary workaround until SciPy's `pinv()` supports sparse matrices.

ageron committed:
Add docstrings for the new sparse svd and pinv functions

ogrisel commented on sklearn/random_projection.py:
```diff
@@ -599,6 +716,12 @@ class SparseRandomProjection(BaseRandomProjection):
 
+    fit_inverse_transform : bool, default=False
+        Learn the inverse transform by computing the pseudo-inverse of the
+        components during fit. Note that the pseudo-inverse is a dense array.
```
```suggestion
        components during fit. Note that the pseudo-inverse is always a
        dense array even if the training data was sparse. This means that
        it might be necessary to call `inverse_transform` on small batch
        of samples at a time to avoid exhausting the available memory on
        the host.
```

ogrisel reviewed (approved):
LGTM. Just a few minor improvements:

ageron committed:
Clarify docstrings as suggested by @ogrisel

ageron commented:
Done

ageron committed:
Merge branch 'main' into rnd_proj_inverse

ageron committed:
Add missing newline in whats_new/v1.1.rst

ageron committed:
Add axis=0 to np.flip()

ageron commented:
Yikes, something's weird: I'm getting no errors in my tests locally, but it's failing in CI. I'll have to investigate this. I'm thinking it might be an error that happens occasionally, if the first k//2 components returned by the first call to `scipy.sparse.linalg.svds()` are not compatible with the k - k//2 components returned by the second call. If that's the case, I don't really see a way to fix that. Any ideas? If there's no solution, I'll just remove `inverse_transform()` for the sparse case. All this hard work for nothing! 😭

ageron committed:
Fix _svd_for_sparse_matrix and remove comp @ pinv(comp) = eye test

ageron commented:
Okay, I've rewritten `_svd_for_sparse_matrix()`, and everything seems to work fine now. Instead of calling `svds()` twice, the code just calls it once. To workaround the fact that `svds` can only return `min(a.shape) - 1` components (missing one), the code now adds an extra row or column full of zeros and crops the `U` or `Vt` matrix appropriately.

Note that I removed the test `assert_array_almost_equal(random_projection.components_ @ random_projection.inverse_components_, np.eye(random_projection.n_components))`. Indeed, a matrix multiplied by its pseudo-inverse is not always equal to the identity matrix, for example:

```python
>>> import numpy as np
>>> from scipy.linalg import pinv
>>> a = np.array([[1, 2, 0], [2, 4, 0]])
>>> a @ pinv(a)
array([[0.2, 0.4],
       [0.4, 0.8]])
```


>>> a = np.array([[1, 2, 0], [2, 4, 0]])
>>> a @ np.linalg.pinv(a)
array([[0.2, 0.4],
       [0.4, 0.8]])

ageron committed:
Add test that inverse_components_ is equal to pinv(inverse_components_)

ageron committed:
Apply black formatting

ageron committed:
Use scipy.linalg.pinv rather than np.linalg.pinv

ageron committed:
Make sparse svd support all cases where n_rows < n_cols, or n_cols < n_rows, or n_cols == n_rows and update tests

ageron committed:
Merge branch 'main' into rnd_proj_inverse

ageron committed:
Black format

ageron commented:
Hi @ogrisel, I just merged main, fixed a couple minor conflicts, it's looking good to me.

jeremiedbb committed:
Update v1.1.rst

jeremiedbb commented on sklearn/random_projection.py:
```diff
@@ -437,6 +530,27 @@ def _n_features_out(self):
         """
         return self.n_components
 
+    @available_if(lambda self: self.fit_inverse_transform)
+    def inverse_transform(self, X):
+        """Project data back to its original space.
```
I don't understand why only exposing `inverse_transform` if the inverse were computed during fit. We could just compute the inverse components if necessary from the `components_` attribute.

ageron replied:
That's what I initially did, if I recall correctly, but I was asked to do this instead, I think it was because that's the pattern used in other classes. I'm not sure what the original rationale was, but perhaps it's because some people would be surprised if the method is available but super slow upon the first call only. It seems reasonable to hide it, especially since it's probably not going to be called very often, so if you really want it, just set `fit_inverse_transform=True`.

jeremiedbb replied:
I don't mean to always do the computation of the inverse components in `inverse_transform`. What I proposed was to keep the `fit_inverse_transform` (not sure about the name btw) parameter and always expose `inverse_transform`. If the param is set to True, the inverse components are computed during fit and stored. If it's set to false they are computed during inverse_transform and not stored.

This parameter only exists in KernelPCA but I'm not sure the use case is really the same. We don't necessarily have to follow the same pattern.

Let's see what others think about that.

jeremiedbb replied:
After an irl discussion with @ogrisel, we agreed on always exposing the `inverse_transformed` method.

jeremiedbb commented on sklearn/random_projection.py:
```diff
@@ -437,6 +530,27 @@ def _n_features_out(self):
+        Returns
+        -------
+        X_original : ndarray of shape (n_samples, n_features)
+            Transformed data.
```
```suggestion
            Reconstructed data.
```

jeremiedbb commented on sklearn/tests/test_random_projection.py:
```diff
@@ -377,6 +378,51 @@ def test_random_projection_feature_names_out(random_projection_cls):
     assert_array_equal(names_out, expected_names_out)
 
 
+@pytest.mark.parametrize("random_state", range(50))
+@pytest.mark.parametrize("n_rows", (2, 9, 10, 11, 1000))
+@pytest.mark.parametrize("n_cols", (2, 9, 10, 11, 1000))
+@pytest.mark.parametrize("random_projection_cls", all_RandomProjection)
```
Wow that makes a lot of tests ! 2500, right ?

We usually don't parametrize tests by seed. It would make the test suite take way too much time. We're thinking about setting up a mechanism to test for different seeds from time to time, but it's another matter.

In the mean time let's just set random_state=0.

ageron replied:
I added this because one test actually passed although there was a bug for some random seeds. IIRC, it was because the random matrix had some empty rows in some cases. But okay, I'll set `random_state=0`.

jeremiedbb replied:
A agree it's a good thing to test against many seeds to check that it's not passing by luck. We are setting up a way to run the tests with a different seed each time in https://github.com/scikit-learn/scikit-learn/pull/22749.

jeremiedbb commented on sklearn/tests/test_random_projection.py:
```diff
@@ -377,6 +378,51 @@ def test_random_projection_feature_names_out(random_projection_cls):
 
 
+@pytest.mark.parametrize("random_state", range(50))
+@pytest.mark.parametrize("n_rows", (2, 9, 10, 11, 1000))
+@pytest.mark.parametrize("n_cols", (2, 9, 10, 11, 1000))
```
Let's use `n_samples`, `n_features`

ageron replied:
Sounds good, I'll change that

jeremiedbb commented on sklearn/tests/test_random_projection.py:
```diff
@@ -377,6 +378,51 @@ def test_random_projection_feature_names_out(random_projection_cls):
+        random_projection = random_projection_cls(
+            n_components=n_components, fit_inverse_transform=True
+        )
+        if n_cols < n_components:
+            with pytest.warns(
+                DataDimensionalityWarning,
+                match="The number of components is higher than the number of features",
+            ):
+                projected = random_projection.fit_transform(data)
+        else:
+            projected = random_projection.fit_transform(data)
```
I think it would be clearer to filter the warning
```py
with warnings.catch_warnings():
    warnings.filterwarnings(
        "ignore",
        message="The number of components is higher than the number of features",
        category=DataDimensionalityWarning)
    projected = random_projection.fit_transform(data)
```

ageron replied:
Ok, I'll change that.

jeremiedbb commented on sklearn/tests/test_random_projection.py:
```diff
@@ -377,6 +378,51 @@ def test_random_projection_feature_names_out(random_projection_cls):
+        projected_again = random_projection.transform(projected_back)
+        if hasattr(projected, "toarray"):
+            projected = projected.toarray()
+        assert_array_almost_equal(projected, projected_again)
```
please use `assert_allclose`

ageron replied:
Sounds good, I'll do that

ageron replied:
Hmm, for some reason it passes with `assert_array_almost_equal()` but not with `assert_allclose()`, I'm not sure why. Tbh I don't have much time right now, I'm trying to finish writing the 3rd edition of my book, so if a kind soul wants to investigate this point, that would be great. What is the difference between `assert_array_almost_equal()` and `assert_allclose()`? May the difference be about sparse vs dense?

jeremiedbb replied:
The difference is that `assert_array_almost_equal` checks the absolute difference between 2 arrays while `assert_allclose` checks the relative difference. We don't want the quality of the test to depend on the magnitude of the arrays.

>  Tbh I don't have much time right now

Don't worry, I'll look into it.

jeremiedbb replied:
The issue is that for the SparseRandomProjection, the transformed data is expected to be sparse and thus have true zeros. When you project back and project again, these zeros can become a very small non-zero value. The relative difference between 0 and any non-zero value is always 100%, thus the failure. We need to set, in addition to the relative tol, an absolute tol. Absolute tol can be seen as "any value below abs_tol is considered to be 0".

I directly pushed some changes in the test

jeremiedbb reviewed (commented):
Thanks @ageron !

Since the inverted components are always dense, why do we bother to implement a pinv for sparse ? Densify the array first and call numpy's pinv could (maybe ?) be more efficient. I understand that the goal was to mitigate the memory usage but we are creating the dense inverse components array anyway.

jeremiedbb committed:
Merge branch 'main' into rnd_proj_inverse

ageron committed:
Update doc/modules/random_projection.rst

Co-authored-by: Jérémie du Boisberranger <34657725+jeremiedbb@users.noreply.github.com>

ageron committed:
Replace "Transformed data" with "Reconstructed data"

Co-authored-by: Jérémie du Boisberranger <34657725+jeremiedbb@users.noreply.github.com>

ageron commented on doc/modules/random_projection.rst:
```diff
@@ -160,3 +160,49 @@ projection transformer::
+  >>> np.allclose(X_new, X_new_again)
+  True
+
+If ``components_`` is dense, then the pseudo-inverse is computed using
+``scipy.linalg.pinv``, which is based on SVD decomposition. If ``components_``
+is sparse, then the pseudo-inverse is computed using an implementation of
+``pinv`` based on ``scipy.sparse.linalg.svds``.
```
Ok, I'll remove this

ageron commented:
> Thanks @ageron !
> 
> Since the inverted components are always dense, why do we bother to implement a pinv for sparse ? Densify the array first and call numpy's pinv could (maybe ?) be more efficient. I understand that the goal was to mitigate the memory usage but we are creating the dense inverse components array anyway.

My pleasure @jeremiedbb !

I think there were two reasons: (1) as you pointed out, it saves saves memory usage (probably halves it), and (2) someone (perhaps @ogrisel ? I can't recall) told me that it may be useful in other places until scipy offers the same functionality, which I believe is on their roadmap.

ageron committed:
Merge branch 'main' into rnd_proj_inverse

ageron committed:
Remove technical detail about how the pseudo-inverse is computed

ageron committed:
Unrelated but flake8 was complaining that a double empty line was missing

ageron committed:
Do not parametrize test for random_state, rename n_cols and n_rows to n_samples and n_features, clarify filter warning

jeremiedbb committed:
take true zeros into account for assert_allclose

jeremiedbb committed:
black

jeremiedbb commented:
> Tbh I don't have much time right now, I'm trying to finish writing the 3rd edition of my book, so if a kind soul wants to investigate this point, that would be great.

If you don't have much time I can directly push these changes if you want.

ageron commented:
Hi @jeremiedbb ,
Thanks for reviewing, and for offering to push the last changes, I really appreciate it. I understand the trade-off about the sparse pinv. There was a bit of a gambler's fallacy on my part: I spent so much time working on the sparse pinv that I was relunctant to just drop it! But not worries, please feel free to delete it.

jeremiedbb committed:
always expose inverse_transform + densify instead of sparse pinv

jeremiedbb committed:
black

jeremiedbb committed:
lint

jeremiedbb committed:
update doc

jeremiedbb committed:
cln

jeremiedbb committed:
updatewhat's new

jeremiedbb committed:
cln

jeremiedbb commented:
>  I spent so much time working on the sparse pinv that I was relunctant to just drop it!

I understand the feeling :)

jeremiedbb committed:
increase coverage

jeremiedbb reviewed (approved):
LGTM

ogrisel committed:
Merge branch 'main' into rnd_proj_inverse

ogrisel committed:
Use the global_random_seed fixture in the new test

ogrisel reviewed (approved):
LGTM again, including the changes suggested and implemented by @jeremiedbb.

Sorry for the back and forth review @ageron :)

I just pushed a merge with main and a new commit to leverage the newly introduced `global_random_seed` fixture in the new test to keep it deterministic while making sure this test is not seed sensitive.

Will merge when green.

jeremiedbb commented:
Thanks @ageron !

ageron commented:
Thanks @ogrisel, @jeremiedbb and @glemaitre for the thorough review, I'm impressed by how pro and helpful you all are, it's no wonder Scikit-Learn is so good! 👍
